torch/lowrank.py

```diff
@@ -0,0 +1,249 @@
+              will be useful for huge sparse matrices that
+              ``torch.svd`` cannot handle.
+
+    Arguments::
+        A (Tensor): the input tensor of size :math:`(*, m, n)`
+
+        q (int, optional): a slightly overestimated rank of A.
+
+        niter (int, optional): the number of subspace iterations to
+                               conduct; niter must be a nonnegative
+                               integer, and defaults to 2
+
+        M (Tensor, optional): the input tensor's mean of size
+                              :math:`(*, 1, n)`.
+
+    References::
+        - Nathan Halko, Per-Gunnar Martinsson, and Joel Tropp, Finding
+          structure with randomness: probabilistic algorithms for
+          constructing approximate matrix decompositions,
+          arXiv:0909.4061 [math.NA; math.PR], 2009 (available at
+          `arXiv <http://arxiv.org/abs/0909.4061>`_).
+
+    """
+    m, n = A.shape[-2:]
+    matmul = _utils.get_matmul(A)
+    if M is None:
+        M_t = None
+    else:
+        matmul2 = _utils.get_matmul(M)
+        M_t = _utils.transpose(M)
+    A_t = _utils.transpose(A)
+
+    # Algorithm 5.1 in Halko et al 2009, slightly modified to reduce
+    # the number conjugate and transpose operations
+    if m < n:
+        # computing the SVD approximation of a transpose in order to
+        # keep B shape minimal
+        Q = get_approximate_basis(A_t, q, niter=niter, M=M_t)
+        Q_c = _utils.conjugate(Q)
+        if M is None:
+            B_t = matmul(A, Q_c)
+        else:
+            B_t = matmul(A, Q_c) - matmul2(M, Q_c)
+        U, S, V = torch.svd(B_t)
+        V = Q.matmul(V)
+    else:
+        Q = get_approximate_basis(A, q, niter=niter, M=M)
+        Q_c = _utils.conjugate(Q)
+        if M is None:
+            B = matmul(A_t, Q_c)
+        else:
+            B = matmul(A_t, Q_c) - matmul2(M_t, Q_c)
+        U, S, V = torch.svd(_utils.transpose(B))
+        U = Q.matmul(U)
+
+    return U, S, V
+
+
+def pca(A, q=None, center=True, niter=2):
+    r"""Performs Principal Component Analysis (PCA) on a low-rank matrix,
+    batches of such matrices, or sparse matrix.
+
+    This function returns a namedtuple ``(U, S, V)`` which is the
+    nearly optimal approximation of a singular value decomposition of
+    a centered matrix :math:`A` such that :math:`A = U diag(S) V^T`.
+
+    .. note:: The relation of ``(U, S, V)`` to PCA is as follows:
+
+                - :math:`A` is a data matrix with ``m`` samples and
+                  ``n`` features
+
+                - the :math:`V` columns represent the principal directions
+
+                - :math:`S ** 2 / (m - 1)` contains the eigenvalues of
+                  :math:`A^T A / (m - 1)` which is the covariance of
+                  ``A`` when ``center=True`` is provided.
+
+                - ``matmul(A, V[:, :k])`` projects data to the first k
+                  principal components
```

Comment:
I disagree. Since A is mxn, where m is the number of samples, then the covariance matrix is actually AA^T, not A^T A. Hence if A = U S V^T, then it is actually U, the left eigenvectors, that represent the principal components.

Reply:
Good catch on the typo :) 

If `A` is a `m x n` matrix, centered, with `m` representing the number of samples, the covariance matrix `C` is `m x m`, and so is `C = A A^T/(m-1)`.

If the (`k`-truncated) singular value decomposition is `A = USV^T`, then `C = U S^2 U^T /(m-1)`. `U` is `m x k` and gives the eigenvectors of the covariance matrix. The projection of the data  `A` onto the eigenspace of `C` is thus the `m x k` matrix `AV`.

The principal directions are the eigenvectors, i.e. `U`. The principal components of the data are the projection of the data, i.e. `AV`.

Hopefully we agree now :)

Reply:
nit: `AU` =)

Reply:
Oooh, I am very sorry, my mistake. The original description is actually correct. My apologies, @pearu, @vincentqb, was super tired :)

Reply:
Confused the covariance matrix the with kernel matrix, happens =)

Reply:
Good to check anyway :)